### PR TITLE
[cli] Add two-factor authentication to login

### PR DESCRIPTION
**Pull request by wschurman**
# Why

This adds CLI support for the upcoming two-factor authentication feature.

## Feature overview: 

Users are able to configure a set of devices to act as second factor authentication mechanisms. There are two kinds of devices: Authenticator TOTP and SMS. A user has the option to have a "primary" device. A user is also given backup codes during 2FA setup.

Currently there is no public way to set up 2FA since it is still in active development.

# How

See inline docblocks.

# Test Plan

Start server locally with https://github.com/expo/universe/pull/5733 patched.

Set up and confirm 2FA for my user in GraphiQL, adding a default authenticator and backup SMS.

```
export EXPO_LOCAL=1

> yarn expo login --username wschurman2 --password <password_in_plain_text_here> --otp 478045
Success. You are now logged in as wschurman2.

> yarn expo logout
Logged out

> yarn expo login
? Username/Email Address: wschurman2
? Password: [hidden]
One-time password from authenticator required.
? One-time Password or Backup Code (press enter for other options): 134243
Success. You are now logged in as wschurman2.

> yarn expo logout
Logged out

> yarn expo login
? Username/Email Address: wschurman2
? Password: [hidden]
One-time password from authenticator required.
? One-time Password or Backup Code (press enter for other options): <press enter, menu showing all devices is shown in next step>
? Select a device to send a one-time password: Phone ending in **74
? One-time Password or Backup Code (press enter to cancel): 769422
Success. You are now logged in as wschurman2.

> yarn expo login
? Username/Email Address: wschurman2
? Password: [hidden]
One-time password from authenticator required.
? One-time Password or Backup Code (press enter for other options):
? Select a device to send a one-time password: Cancel
User cancelled login
```

Also, run the new integration test.

### CODE REVIEW

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:135]:
```suggestion
 * for a new OTP will be sent to that method. Then, prompt for the OTP, and retry the user login.
```
```suggestion
 * for a new  will be sent to that method. Then, prompt for the OTP, and retry the user login.
```

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:145]: Since these error messages are user-facing, try to phrase this to be more directed at the user e.g.
```suggestion
    throw new Error('There are no other second-factor devices set up for you');
```
```suggestion
    throw new Error('No other second factor devices set up for user');
```

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:160]: This prompt looks OK but might want to change the phrasing if we change the masking format (comment is on the server-side PR).

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:179]:
```suggestion
    message: 'Select a second-factor device:',
```

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:186]: This is a little more user-facing by not referring to "user in the third person:
```suggestion
    throw new Error('Cancelled login');
```

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:232]: `metadata` is already typed in the param list, most of these types seem unnecessary

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:246]: I'm not familiar with this codebase but I would double check that `console.log` is the established way to write to the terminal instead of via some other logging API

File: packages/expo-cli/src/accounts.ts
Comment (ide) [inline packages/expo-cli/src/accounts.ts:263]:
```suggestion
    throw new Error('Cancelled login');
```
```suggestion
    throw new Error('User cancelled login');
```

**Review by ide — CHANGES_REQUESTED** [review]
@fson Could you please review this too? We'll also need 2FA support with EAS CLI and will need to support the same type of login flows and error handling.

**Comment by ide** [comment]
@EvanBacon Could you take a look at the "Test Plan" section in the PR description and see if the flow and the wording sounds good, or if you think there are some tweaks we should make to it?

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:232]: Ah, good call, this was leftover from refactoring.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:246]: As far as I can tell via grepping it is the correct method.

File: packages/expo-cli/src/accounts.ts
Comment (brentvatne) [inline packages/expo-cli/src/accounts.ts:229]: should we include the stringified metadata in the message so we have more context if a user reports this happening?

**Review by brentvatne — COMMENTED** [review]
looks good! excited to have 2fa soon

- this will need to be updated if https://github.com/expo/expo-cli/pull/2605 lands first
- we should use [`log`](https://github.com/expo/expo-cli/blob/90bab4a0c6ab8f39a5dccf71f395e0cca7b8f626/packages/expo-cli/src/log.ts#L91-L95) instead of `console.log` directly - this just makes sure that we don't end up interrupting any progress bars / spinners that might be active. if you can change other `console.log`s that were leftover in this file prior to your changes that'd be great too

**Comment by ide** [comment]
Generally speaking this looks good to me and consider this a soft approval. Recusing myself from review so we don't run the risk of no single reviewer looking at this carefully. @fson please look this over (and we'll need similar code for EAS CLI); @EvanBacon please review whether the DX / workflow seems good to you (shown in the test plan) or if @brentvatne is happy with it that works for me.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:229]: it could be a fairly large object but I think something along these lines is a good idea. I'll dig into how we do this in other places and update this.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [inline packages/expo-cli/src/accounts.ts:150]: hmm, this logic is incorrect now. This should be `hasAuthenticatorSecondFactorDevice` and the condition should be inverted.

**Comment by wschurman** [comment]
> we should use log instead of console.log directly - this just makes sure that we don't end up interrupting any progress bars / spinners that might be active. if you can change other console.logs that were leftover in this file prior to your changes that'd be great too

I'll make this change in a follow-up PR to keep the scope of this one as narrow as possible.

**Comment by wschurman** [comment]
Still looking for review on this from @fson and/or @EvanBacon. The two branches for disallowing console and enforcing no lint warnings in CI were merged here because github is bad at stacked PRs and should feel bad. I'll un-merge them before squash landing and re-PR those branches and land them separately.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:118]:
```suggestion
    cancelBehavior === 'cancel' ? `press ${log.chalk.bold('Enter')} to cancel` : `press ${log.chalk.bold('Enter')} for other options`;
```

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:125]: Use `prompts` instead of `prompt`, prompts uses a substantially smaller prompting package.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:137]: maybe this method could be split up so the aggregation for the question and the actual interaction/api call are in two different methods, then you should be able to easily write a unit test for the aggregation.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:230]: if this is not an unexpected error, then be sure the error message looks actionable so a user knows what to do next.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:238]: Since we know the cause of the problem, perhaps we could format this more like a warning rather than an error. For example, when someone runs `expo start --config` instead of throwing an invalid type, we warn how to use flag and exit with non-zero. https://github.com/expo/expo-cli/blob/50f98f66a52cb3371aa580d9d504d465ce8558a6/packages/expo-cli/src/exp.ts#L418-L422 the lack of generic error and a stack trace helps people know that CLI isn't buggy.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:247]:
```suggestion
    log(`One-time password from authenticator required.`);
```

This helps for deferring logs when something with higher priority is currently in use.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:257]: as above, so below :D because the user cancelled, this is expected behavior and not an error. We should log something with default styles and exit the process.

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:122]:
```suggestion
    message: `One-time password or backup code (${enterMessage}):`,
```

File: packages/expo-cli/src/accounts.ts
Comment (EvanBacon) [inline packages/expo-cli/src/accounts.ts:145]:
```suggestion
    log.warn(`You need to set up two factor authentication before you can use a OTP. ${log.chalk.dim(learnMore('<link to docs or fyi.expo>'))}`);
    process.exit(1);
```
Maybe create an FYI or link to docs about how to set up a device.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:230]: This is an unexpected error. It'd be thrown when someone in the future makes an erroneous (non-backwards-compatible) change on the server, and when that happens we want to find out about it in sentry. I'm assuming throwing an invariant will go to our sentry but if there's a better way to log this let me know.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:238]: This is also a constraint that is easy to express on the server but hard to express over an API. Essentially, when an SMS is automatically sent that means that there is a primary device and it is an SMS device. The invariant here is to appease TypeScript to ensure that this is non-null for use.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:145]: This error occurs when the user only has one second factor device set up and chooses the option (`press Enter for more options`) in the prompt. We still want them to be shown this option though even if they only have one second factor device as they could be in a state where they have a second one that is in the process of being setup but is not yet certified/confirmed, in which case this error will help them debug their issue. I'll alter the wording a bit.

Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:145]: I'll make a task to add in a link once we have documentation.

**Review by fson — CHANGES_REQUESTED** [review]
Nice work, especially thank you for being thoughtful about different error conditions and how to handle them.

I think we need to avoid calling `process.exit()` in the login prompt functions as this can interfere with the dev server / Expo Dev Tools, which can also trigger the login prompt. Otherwise everything seems good and well tested, so consider this PR approved once `process.exit` is fixed.

File: packages/expo-cli/src/accounts.ts
Comment (wschurman) [reply packages/expo-cli/src/accounts.ts:257]: Sounds good! Will proceed with `CommandError`.

cc @EvanBacon for alignment as this contradicts the requested change when an error was previously thrown here: https://github.com/expo/expo-cli/pull/2581#discussion_r491070772

Comment (EvanBacon) [reply packages/expo-cli/src/accounts.ts]: I suppose you can put better styling here then https://github.com/expo/expo-cli/blob/ff0014cc5e48acb6c86eeb0de02ba54a38427ee6/packages/expo-cli/src/exp.ts#L308 really no other tools should be calling into a CLI unless they're forking a new process anyways.

Comment (wschurman) [reply packages/expo-cli/src/accounts.ts]: Looks like the styling is already the same as `log.warn` (orange in my terminal) so I think we're good to go.

**Review by fson — APPROVED** [review]

**Pull request merged**
